### PR TITLE
PSA Proposed appendix discussing multi-pipeline PSA impelementation i…

### PR DESCRIPTION
**Pull request by jafingerhut**
…ssues

### CODE REVIEW

**Comment by jafingerhut** [comment]
@anirudhSK @jnfoster Given your participation in the issue that led to this proposed addition to the PSA spec, I wanted to point this out in case you wanted to read it over and comment.

File: p4-16/psa/PSA.mdk [hunk 1]
```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
...
+start processing a new packet once every clock cycle, and finish a
+packet every clock cycle.  The latency might be hundreds of clock
+cycles from starting a packet until it is complete.  P4 tables in such
+ASICs are typically implemented using logic such as TCAMs and SRAMs.
+TCAM designs can do 1 search per clock cycle.  The lowest area and
+power SRAMs can do 1 read or 1 write per clock cycle.
+
+While there are "multi ported" SRAM designs that can be read and/or
+written multiple times per clock cycle, these have a noticeable
+increase in area and power over the "singled ported" designs that are
```
Comment (anirudhSK) [inline p4-16/psa/PSA.mdk]: Think there's a typo. Should be "single ported".

File: p4-16/psa/PSA.mdk [hunk 2]
```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
...
+limited to 1 access per cycle.  If multi ported TCAM designs even
+exist, they are likely even a higher cost premium of single ported
```
Comment (anirudhSK) [inline p4-16/psa/PSA.mdk]: Had a bit of a hard time parsing this sentence fragment. I suggest we reword it to: "they are likely to have an even higher cost premium (relative to single ported TCAMs) than for SRAMs."

File: p4-16/psa/PSA.mdk [hunk 3]
```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
...
+is via parallelism, by creating multiple copies of the desired TCAM,
+which is a linear increase in area and power (at least).
+
+Due to these issues, if one wishes to create a switch ASIC that
+achieves a packet processing rate of N times the clock rate, e.g. 2
+billion packets per second in a 1 GHz ASIC, the most straightforward
+way to do this is to take advantage of the fact that packet processing
+is (mostly) an embarrasingly parallel
+problem[^EmbarrassinglyParallel], and create a device with N modules,
+each modules processing packets at 1 billion packets per second.
```
Comment (anirudhSK) [inline p4-16/psa/PSA.mdk]: "each modules" => each module. Do you think it might help to use the term pipeline instead of module? Or was your intent to cover both pipeline and NPU-style architectures?

File: p4-16/psa/PSA.mdk [hunk 4]
```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
...
+
+[^EmbarrassinglyParallel]: https://en.wikipedia.org/wiki/Embarrassingly_parallel
+
+A PSA device designed in this way would typically have N ingress
+pipeline modules, plus N egress pipeline modules.  It is common to
```
Comment (anirudhSK) [inline p4-16/psa/PSA.mdk]: This is a bit confusing. The term pipeline wasn't introduced until this point; everything so far was about modules. But now we have pipeline modules. We could either use pipeline stages and pipelines throughout or change it to modules throughout. I personally think pipeline stages and pipelines are clearer.

File: p4-16/psa/PSA.mdk [hunk 5]
```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
...
+It is physically possible in such a device for the control plane to
+install different table entries in the different pipelines, and there
+are use cases where this can help achieve higher scale in number of
+usable table entries.  For example, perhaps you have an ingress table
+with the ingress port as one of the fields in its search key.  If this
+is the case, the packet processing behavior is the same even if table
+entries for port X are installed only in the ingress pipeline that
+processes packets from port X.  Installing that table entry in the
+other pipelines would be an unnecessary use of table space, since it
+could never be matched.  Whether the software of such a PSA device
```
Comment (anirudhSK) [inline p4-16/psa/PSA.mdk]: What is the software here? The control plane or the data plane program?

File: p4-16/psa/PSA.mdk [hunk 6]
```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
...
+enables taking advantage of such space savings is implementation
+dependent.
+
+Regardless of this issue, applying tables in P4 is an embarrassingly
+parallel activity, because as long as table entries are installed
+where they might be matched, each pipeline can operate completely
```
Comment (anirudhSK) [inline p4-16/psa/PSA.mdk]: Minor nit. Suggest replacing "each pipeline" with "pipelines".

**Comment by anirudhSK** [comment]
Andy, Thanks a bunch for writing this up. It's very clear and helpful. I have left a few minor grammar/typo comments in the pull request. My only concern is the use of the word module instead of pipeline/pipeline stage. I am OK with either set of terms, but would prefer that we be consistent throughout the appendix.  Besides that, it looks really good.

File: p4-16/psa/PSA.mdk (hunk 1 above)
Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: Fixed.

File: p4-16/psa/PSA.mdk (hunk 2 above)
Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: Yeah, that was an akward sentence.  I like your suggested replacement better.  Grabbed!

File: p4-16/psa/PSA.mdk (hunk 3 above)
Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: Yes, I will use pipeline throughout, trying to restrict myself on the first use to a footnote that the term should not be taken to mean that a hardware pipeline is the only kind of implementation possible.

File: p4-16/psa/PSA.mdk (hunk 4 above)
Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: Done.

File: p4-16/psa/PSA.mdk (hunk 5 above)
Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: Changed "software" to "device-dependent control plane software" in hopes of clarifying this.

File: p4-16/psa/PSA.mdk (hunk 6 above)
Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: done

File: p4-16/psa/PSA.mdk [hunk 7]
```diff
@@ -2576,3 +2576,127 @@ Certainly if there are long durations of time when egress decides to
...
+ASICs are typically implemented using logic such as TCAMs and SRAMs.
+TCAM designs can do 1 search per clock cycle.  The lowest area and
+power SRAMs can do 1 read or 1 write per clock cycle.
+
+While there are "multi ported" SRAM designs that can be read and/or
+written multiple times per clock cycle, these have a noticeable
+increase in area and power over the "singled ported" designs that are
+limited to 1 access per cycle.  If multi ported TCAM designs even
+exist, they are likely even a higher cost premium of single ported
+TCAMs than for SRAMs.  They typical way to achieve higher search rates
```
Comment (ChrisDodd) [inline p4-16/psa/PSA.mdk]: "They typical" -> "The typical".

Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: Good catch.  Fixed.

File: p4-16/psa/PSA.mdk [hunk 8]
```diff
@@ -2576,3 +2576,135 @@ Certainly if there are long durations of time when egress decides to
...
+billion packets per second in a 1 GHz ASIC, the most straightforward
+way to do this is to take advantage of the fact that packet processing
+is (mostly) an embarrasingly parallel
+problem[^EmbarrassinglyParallel], and create a device with N
+pipelines[^PipelineNote], each pipeline processing packets at 1
+billion packets per second.
+
+[^EmbarrassinglyParallel]: https://en.wikipedia.org/wiki/Embarrassingly_parallel
+
+[^PipelineNote]: Here and elsewhere in P4 specification documents, the term "pipeline" is often used for such a thing.  This is by now traditional in P4 specifications, and although we will not try to change that here, note that there are other reasonable hardware designs that can accomplish this goal, e.g. a collection of CPU cores running in parallel, each processing different packets.
```
Comment (cc10512) [inline p4-16/psa/PSA.mdk]: "used for such a thing"? Maybe: used to refer to the sequence of blocks operating on the packet?

Comment (jafingerhut) [reply p4-16/psa/PSA.mdk]: Well "sequence" has similar connotations as a pipeline.  P4 implementations need not be implemented as a pipeline.  I am pretty sure that this name is used because of the RMT architecture being a hardware pipeline design.

**Comment by jafingerhut** [comment]
@cc10512 Attempted to address your comments, and am going to go ahead and merge this in now.  Sorry if I am jumping the gun there -- please do take a look at the changes I made to try to address your comments and let me know / make a new PR if it misses the mark.
